sys/net/application_layer/coapfileserver/coapfileserver.c

```diff
@@ -0,0 +1,401 @@
+ * etc., defaulting to 5.03 Internal Server Error) */
+static size_t coapfileserver_errno_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, int err);
+/** Return true if path/name is a directory.
+ *
+ * Currently assembling the values in a buffer, this could be done without a
+ * buffer if an openat / vfs_openat was present.
+ * */
+static bool entry_is_dir(char *path, char *name);
+
+ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *ctx) {
+    struct coapfileserver_entry *entry = (struct coapfileserver_entry *)ctx;
+    struct requestdata request;
+    request.etag_sent = false;
+    request.blocknum2 = 0;
+    request.szx2 = CONFIG_NANOCOAP_BLOCK_SIZE_EXP_MAX;
+    /** Index in request.namebuf. Must not point at the last entry as that will be
+     * zeroed to get a 0-terminated string. */
+    size_t namelength = 0;
+    bool trailing_slash = false;
+    /** If no path component comes along at all, it'll count as a trailing
+     * slash no matter the trailing_slash value */
+    bool any_component = false;
+    uint8_t errorcode = COAP_CODE_INTERNAL_SERVER_ERROR;
+
+    uint8_t strip_remaining = entry->strip_path;
+    namelength += request.namebuf - strncpy(request.namebuf, entry->nameprefix, sizeof(request.namebuf) - 1);
+
+    coap_optpos_t opt;
+    bool is_first = true;
+    uint8_t *value;
+    ssize_t optlen;
+    while ((optlen = coap_opt_get_next(pdu, &opt, &value, is_first)) !=
+            -ENOENT)
+    {
+        if (optlen < 0) {
+            errorcode = COAP_CODE_BAD_REQUEST;
+            goto error;
+        }
+
+        is_first = false;
+        switch (opt.opt_num) {
+            case COAP_OPT_URI_PATH:
+                if (strip_remaining != 0) {
+                    strip_remaining -= 1;
+                    continue;
+                }
+                if (trailing_slash) {
+                    errorcode = COAP_CODE_BAD_REQUEST;
+                    goto error;
+                }
+                any_component = true;
+                if (optlen == 0) {
+                    trailing_slash = true;
+                    continue;
+                }
+                if (memchr(value, '0', optlen) != NULL ||
+                        memchr(value, '/', optlen) != NULL) {
+                    /* Path can not be expressed in the file system */
+                    errorcode = COAP_CODE_PATH_NOT_FOUND;
+                    goto error;
+                }
+                size_t newlength = namelength + 1 + optlen;
+                if (newlength > sizeof(request.namebuf) - 1) {
+                    /* Path too long, therefore can't exist in this mapping */
+                    errorcode = COAP_CODE_PATH_NOT_FOUND;
+                    goto error;
+                }
+                request.namebuf[namelength] = '/';
+                memcpy(&request.namebuf[namelength] + 1, value, optlen);
+                namelength = newlength;
```

Comment:
Is there really no way in GCoap to provide the handler with the request URL, so that the handler doesn't have to parse the coap packet itself?